libraries/chain/abi_serializer.cpp

```diff
@@ -32,15 +32,33 @@ namespace eosio { namespace chain {
    template <typename T>
    auto pack_unpack() {
       return std::make_pair<abi_serializer::unpack_function, abi_serializer::pack_function>(
-         []( fc::datastream<const char*>& stream, bool is_array, bool is_optional) -> fc::variant  {
-            if( is_array )
+         []( fc::datastream<const char*>& stream, bool is_array, bool is_optional, int fixed_array_size) -> fc::variant  {
```

Comment:
Is this necessary? Can we pack/unpack fixed_array same as array?

Reply:
hi @heifner the real issue behind is that if users don't use EOSLIB_SERIALIZE in their struct definition, boost::pfr::for_each_field will be used to serialize struct members. In this case, an array member of size N will be break down into N elements, then there's no way to tell the size of the array from the datastream.

Reply:
hi @heifner,

This is an example of contract that doesn't work:

```
#include <eosiolib/eosio.hpp>
#include <eosiolib/action.hpp>
using namespace eosio;
class hello : public eosio::contract {
  public:
      using contract::contract;

      /// @abi table
      struct data {
         uint64_t     key;
         uint64_t     int64array[2];
         uint8_t      chksum = 77;
         uint8_t      chksum2 = 99;
         uint64_t     primary_key() const { return key; }
      };

      typedef eosio::multi_index<N(data), data> data_table;

      /// @abi action
      void setkey(uint64_t scope, uint64_t key) {
         data_table table(_self, scope);
         auto iter = table.find(key);
         if (iter == table.end()) {
            table.emplace(_self, [&](auto &v) {
               v.key = key;
               v.int64array[0] = 10;
               v.int64array[1] = 20;
            });
         }
      }

};
EOSIO_ABI(hello, (setkey))
```

abi file:
```
{
  "____comment": "This file was generated by eosio-abigen. DO NOT EDIT - 2018-08-17T01:53:28",
  "version": "eosio::abi/1.0",
  "types": [],
  "structs": [{
      "name": "data",
      "base": "",
      "fields": [{
          "name": "key",
          "type": "uint64"
        },{
          "name": "int64array",
          "type": "uint64[]"
        },{
          "name": "chksum",
          "type": "uint8"
        },{
          "name": "chksum2",
          "type": "uint8"
        }
      ]
    },{
      "name": "setkey",
      "base": "",
      "fields": [{
          "name": "scope",
          "type": "uint64"
        },{
          "name": "key",
          "type": "uint64"
        }
      ]
    }
  ],
  "actions": [{
      "name": "setkey",
      "type": "setkey",
      "ricardian_contract": ""
    }
  ],
  "tables": [{
      "name": "data",
      "index_type": "i64",
      "key_names": [
        "key"
      ],
      "key_types": [
        "uint64"
      ],
      "type": "data"
    }
  ],
  "ricardian_clauses": [{
      "id": "Warranty",
      "body": "WARRANTY. The invoker of the contract action shall uphold its Obligations under this Contract in a timely and workmanlike manner, using knowledge and recommendations for performing the services which meet generally acceptable standards set forth by EOS.IO Blockchain Block Producers.\n\n"
    },{
      "id": "Default",
      "body": "DEFAULT. The occurrence of any of the following shall constitute a material default under this Contract: \n\n"
    },{
      "id": "Remedies",
      "body": "REMEDIES. In addition to any and all other rights a party may have available according to law, if a party defaults by failing to substantially perform any provision, term or condition of this Contract, the other party may terminate the Contract by providing written notice to the defaulting party. This notice shall describe with sufficient detail the nature of the default. The party receiving such notice shall promptly be removed from being a Block Producer and this Contract shall be automatically terminated. \n  \n"
    },{
      "id": "Force Majeure",
      "body": "FORCE MAJEURE. If performance of this Contract or any obligation under this Contract is prevented, restricted, or interfered with by causes beyond either party's reasonable control (\"Force Majeure\"), and if the party unable to carry out its obligations gives the other party prompt written notice of such event, then the obligations of the party invoking this provision shall be suspended to the extent necessary by such event. The term Force Majeure shall include, without limitation, acts of God, fire, explosion, vandalism, storm or other similar occurrence, orders or acts of military or civil authority, or by national emergencies, insurrections, riots, or wars, or strikes, lock-outs, work stoppages, or supplier failures. The excused party shall use reasonable efforts under the circumstances to avoid or remove such causes of non-performance and shall proceed to perform with reasonable dispatch whenever such causes are removed or ceased. An act or omission shall be deemed within the reasonable control of a party if committed, omitted, or caused by such party, or its employees, officers, agents, or affiliates. \n  \n"
    },{
      "id": "Dispute Resolution",
      "body": "DISPUTE RESOLUTION. Any controversies or disputes arising out of or relating to this Contract will be resolved by binding arbitration under the default rules set forth by the EOS.IO Blockchain. The arbitrator's award will be final, and judgment may be entered upon it by any court having proper jurisdiction. \n  \n"
    },{
      "id": "Entire Agreement",
      "body": "ENTIRE AGREEMENT. This Contract contains the entire agreement of the parties, and there are no other promises or conditions in any other agreement whether oral or written concerning the subject matter of this Contract. This Contract supersedes any prior written or oral agreements between the parties. \n\n"
    },{
      "id": "Severability",
      "body": "SEVERABILITY. If any provision of this Contract will be held to be invalid or unenforceable for any reason, the remaining provisions will continue to be valid and enforceable. If a court finds that any provision of this Contract is invalid or unenforceable, but that by limiting such provision it would become valid and enforceable, then such provision will be deemed to be written, construed, and enforced as so limited. \n\n"
    },{
      "id": "Amendment",
      "body": "AMENDMENT. This Contract may be modified or amended in writing by mutual agreement between the parties, if the writing is signed by the party obligated under the amendment. \n\n"
    },{
      "id": "Governing Law",
      "body": "GOVERNING LAW. This Contract shall be construed in accordance with the Maxims of Equity. \n\n"
    },{
      "id": "Notice",
      "body": "NOTICE. Any notice or communication required or permitted under this Contract shall be sufficiently given if delivered to a verifiable email address or to such other email address as one party may have publicly furnished in writing, or published on a broadcast contract provided by this blockchain for purposes of providing notices of this type. \n"
    },{
      "id": "Waiver of Contractual Right",
      "body": "WAIVER OF CONTRACTUAL RIGHT. The failure of either party to enforce any provision of this Contract shall not be construed as a waiver or limitation of that party's right to subsequently enforce and compel strict compliance with every provision of this Contract. \n\n"
    },{
      "id": "Arbitrator's Fees to Prevailing Party",
      "body": "ARBITRATOR'S FEES TO PREVAILING PARTY. In any action arising hereunder or any separate action pertaining to the validity of this Agreement, both sides shall pay half the initial cost of arbitration, and the prevailing party shall be awarded reasonable arbitrator's fees and costs.\n  \n"
    },{
      "id": "Construction and Interpretation",
      "body": "CONSTRUCTION AND INTERPRETATION. The rule requiring construction or interpretation against the drafter is waived. The document shall be deemed as if it were drafted by both parties in a mutual effort. \n  \n"
    }
  ],
  "error_messages": [],
  "abi_extensions": []
}
```

step to produce:
./cleos -v push action a123 setkey '{"scope":1,"key":1}' -p a123
./cleos -v get table a123 1 data

Reply:
Thanks for the example. It is very helpful. Please try out my PR and see if it works for you. Is this  going into a doc or tutorial? We will need to make sure the javascript libs are updated.

Reply:
After discussing this with @larryk85 @tbfleming we are reluctant to merge this PR at this time. Bucky is working on serialization for the new wasmsdk and we don't want to commit to supporting c-style arrays at this time.

In your example, you could use vector or 
```
      struct data {
         uint64_t     key;
         uint64_t     int64array_0;
         uint64_t     int64array_1;
         uint8_t      chksum = 77;
         uint8_t      chksum2 = 99;
         uint64_t     primary_key() const { return key; }
      };
```